mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -286,40 +288,119 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
+    ).flatMap { case (index, (srcIds, dstIdMatrix, scoreMatrix)) =>
+      // to avoid corner case that the number of items is less than recommendation num
+      var col: Int = 0
+      while (col < num && scoreMatrix(0, col) > Double.NegativeInfinity) {
+        col += 1
+      }
+      val row = scoreMatrix.numRows
+      val output = new Array[(Int, Array[(Int, Double)])](row)
       var i = 0
-      val pq = new BoundedPriorityQueue[(Int, Double)](n)(Ordering.by(_._2))
-      srcIter.foreach { case (srcId, srcFactor) =>
-        dstIter.foreach { case (dstId, dstFactor) =>
-          // We use F2jBLAS which is faster than a call to native BLAS for vector dot product
-          val score = BLAS.f2jBLAS.ddot(rank, srcFactor, 1, dstFactor, 1)
-          pq += dstId -> score
+      while (i < row) {
+        val factors = new Array[(Int, Double)](col)
+        var j = 0
+        while (j < col) {
+          factors(j) = (dstIdMatrix(i * num + j), scoreMatrix(i, j))
+          j += 1
         }
-        pq.foreach { case (dstId, score) =>
-          output(i) = (srcId, (dstId, score))
-          i += 1
+        output(i) = (srcIds(i), factors)
+        i += 1
+      }
+     output.toSeq}
+  }
+
+  private def mergeFunc(rateSum: (Array[Int], Array[Int], DenseMatrix),
+                        rate: (Array[Int], Array[Int], DenseMatrix),
+                        num: Int): (Array[Int], Array[Int], DenseMatrix) = {
+    if (rateSum._1 == null) {
+      rate
+    } else {
+      val row = rateSum._3.numRows
+      var i = 0
+      val tempIdMatrix = new Array[Int](row * num)
+      val tempScoreMatrix = Array.fill[Double](row * num)(Double.NegativeInfinity)
+      while (i < row) {
+        var j = 0
+        var sum_index = 0
+        var rate_index = 0
+        val matrixIndex = i * num
+        while (j < num) {
+          if (rate._3(i, rate_index) > rateSum._3(i, sum_index)) {
+            tempIdMatrix(matrixIndex + j) = rate._2(matrixIndex + rate_index)
+            tempScoreMatrix(matrixIndex + j) = rate._3(i, rate_index)
+            rate_index += 1
+          } else {
+            tempIdMatrix(matrixIndex + j) = rateSum._2(matrixIndex + sum_index)
+            tempScoreMatrix(matrixIndex + j) = rateSum._3(i, sum_index)
+            sum_index += 1
+          }
+          j += 1
         }
-        pq.clear()
+        i += 1
       }
-      output.toSeq
+      (rateSum._1, tempIdMatrix, new DenseMatrix(row, num, tempScoreMatrix, true))
     }
-    ratings.topByKey(num)(Ordering.by(_._2))
   }
 
   /**
    * Blockifies features to improve the efficiency of cartesian product
    * TODO: SPARK-20443 - expose blockSize as a param?
    */
-  private def blockify(
-      features: RDD[(Int, Array[Double])],
-      blockSize: Int = 4096): RDD[Seq[(Int, Array[Double])]] = {
+  def blockify(
+      rank: Int,
+      features: RDD[(Int, Array[Double])]): RDD[(Array[Int], DenseMatrix)] = {
+    val blockSize = 2000 // TODO: tune the block size
```

Comment:
So will you add a parameter for this ?

Reply:
Yes, we have another PR to set this value SPARK-20443.
If the blockSize is large enough, it is possible to OOM.  For my test, the blockSize is set from 1000 to 8000, the performance of this PR is better than the master. 
And the performance is about the same for blockSize is 1000 to 8000.